From 3796d7d94ef87a72fd02a0d74256bfc3ae476cdb Mon Sep 17 00:00:00 2001
From: Garrett Brown <themagnificentmrb@gmail.com>
Date: Sun, 2 Oct 2016 10:08:11 -0700
Subject: [PATCH] Add ability to load resources from multiple directories

---
 libretro.c | 29 ++++++++++++++++++++++-------
 libretro.h | 13 +++++++++++++
 2 files changed, 35 insertions(+), 7 deletions(-)

diff --git a/libretro.c b/libretro.c
index ce13fee..03af7d3 100644
--- a/libretro.c
+++ b/libretro.c
@@ -209,6 +209,12 @@ static unsigned btn_map[EC_KEYCOUNT] =
    RETROK_UNKNOWN,   //EC_COLECO2_HASH 151
 };
 
+#ifdef _WIN32
+#define SLASH_CHAR  '\\'
+#else
+#define SLASH_CHAR  '/'
+#endif
+
 
 retro_log_printf_t log_cb;
 static retro_video_refresh_t video_cb;
@@ -428,17 +434,26 @@ static void check_variables(void)
       msx_ym2413_enable = true;
 }
 
+void set_path(char *dest, size_t maxlen, const char *rel_path, const char *fallback)
+{
+   const char *base_dir = NULL;
+
+   struct retro_resource res = { rel_path };
+   if (environ_cb(RETRO_ENVIRONMENT_GET_RESOURCE_DIRECTORY, &res))
+      base_dir = res.base_directory;
+
+   if (!base_dir)
+      base_dir = fallback;
+
+   snprintf(dest, maxlen, "%s%c%s", base_dir, SLASH_CHAR, rel_path);
+}
+
 bool retro_load_game(const struct retro_game_info *info)
 {
    int i;
    char properties_dir[256], machines_dir[256], mediadb_dir[256];
    const char *dir = NULL;
    enum retro_pixel_format fmt = RETRO_PIXEL_FORMAT_RGB565;
-#ifdef _WIN32
-   char slash = '\\';
-#else
-   char slash = '/';
-#endif
 
    if (!environ_cb(RETRO_ENVIRONMENT_SET_PIXEL_FORMAT, &fmt))
    {
@@ -454,8 +469,8 @@ bool retro_load_game(const struct retro_game_info *info)
    else /* Fallback */
       extract_directory(properties_dir, info->path, sizeof(properties_dir));
 
-   snprintf(machines_dir, sizeof(machines_dir), "%s%c%s", properties_dir, slash, "Machines");
-   snprintf(mediadb_dir, sizeof(mediadb_dir), "%s%c%s", properties_dir, slash, "Databases");
+   set_path(machines_dir, sizeof(machines_dir), "Machines", properties_dir);
+   set_path(mediadb_dir, sizeof(mediadb_dir), "Databases", properties_dir);
 
    propertiesSetDirectory(properties_dir, properties_dir);
    machineSetDirectory(machines_dir);
diff --git a/libretro.h b/libretro.h
index 16c274a..2ab79e5 100644
--- a/libretro.h
+++ b/libretro.h
@@ -852,6 +852,19 @@ enum retro_mod
                                             * It can be used by the core for localization purposes.
                                             */
 
+struct retro_resource
+{
+   const char *rel_path;
+   const char *base_directory;
+};
+
+#define RETRO_ENVIRONMENT_GET_RESOURCE_DIRECTORY 99
+                                           /* struct retro_resource * --
+                                            * Interface to acquire resource paths by a relative filename.
+                                            * 'rel_path' should be set to a filename relative to the system directory.
+                                            * 'base_directory' will be set to the base path containing the resource or NULL if unknown.
+                                            */
+
 #define RETRO_MEMDESC_CONST     (1 << 0)   /* The frontend will never change this memory area once retro_load_game has returned. */
 #define RETRO_MEMDESC_BIGENDIAN (1 << 1)   /* The memory area contains big endian data. Default is little endian. */
 #define RETRO_MEMDESC_ALIGN_2   (1 << 16)  /* All memory access in this area is aligned to their own size, or 2, whichever is smaller. */
-- 
2.7.4

